### PR TITLE
[BACKUP] Fix #14976: CLI error improvements for ValueError and AttributeError cases

### LINKED ISSUE #14976: Fix non-actionable code bugs: Backup
These issues are from Telemetry records, and most of them are caused by obvious code bugs. The detail error messages and the corresponding command and CLI version info can be viewed in the [CLI error dashboard](https://msit.powerbi.com/groups/de995b1a-16c1-4cc7-a49b-bc9f90bc6acd/reports/b657544c-ab0a-4306-bc7b-5bedab1bb5a9/ReportSection01812df080cad06afcb4?noSignUpCheck=1power).

As the info of these issues are limited and may be more difficult to identify and solve, we opened a [Teams channel](https://teams.microsoft.com/l/team/19%3a0e3d8b3a0c40468884a92709305d5898%40thread.tacv2/conversations?groupId=863077aa-7638-4956-a60e-43b88698fb04&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47) for command groups owners to sync up the common error types and fixing methods. Feel free to share your experience and ask for help (if any) there.

A brief summary of the code bug errors of BACKUP here:

FailureType | EventCount
-- | --
AttributeError | 1014
ValueError | 529
TypeError | 126
KeyError | 61
UnknownFunctionError | 10
JMESPathTypeError | 4
ArityError | 3
SyntaxError | 1

### PR DESCRIPTION
**Pull request by akshayneema**
**Description**  
Replaced ValueError with ResourceNotFoundError to eliminate traceback from the error shown. Added a few more validation functions in custom_help.py to customize ResourceNotFoundError message. Added some validations/checks to eliminate AttributeError.

**Testing Guide**  

**History Notes**  

[Component Name 1] BREAKING CHANGE: az command a: Make some customer-facing breaking change.  
[Component Name 2] az command b: Add some customer-facing feature.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

**Review by sambitratha — APPROVED** [review]

File: src/azure-cli/azure/cli/command_modules/backup/custom_base.py [hunk 1]
```diff
@@ -197,6 +197,8 @@ def list_protectable_items(cmd, client, resource_group_name, vault_name, workloa
             container = show_container(cmd, container_client, container_name, resource_group_name, vault_name,
                                        "AzureWorkload")
             custom_help.validate_container(container)
+            if isinstance(container, list):
+                raise CLIError("Multiple containers with same Friendly Name found. Please give native names instead.")
```
Comment (fengzhou-msft) [inline src/azure-cli/azure/cli/command_modules/backup/custom_base.py:201]: Please give the option name for native names in the error message to make it more clear for users. @houk-ms do we have a more detailed error type for this case?

Comment (akshayneema) [reply src/azure-cli/azure/cli/command_modules/backup/custom_base.py:201]: There is a similar error raised on line 402 of this file. I just replicated that error.

File: src/azure-cli/azure/cli/command_modules/backup/custom_wl.py [hunk 2]
```diff
@@ -636,6 +639,14 @@ def _get_protected_item_instance(item_type):
 
 
 def _check_map(item_type, item_type_map):
+    if item_type is None:
+        if item_type_map == workload_type_map:
+            raise RequiredArgumentMissingError("Workload type missing. Please enter a valid workload type.")
```
Comment (houk-ms) [inline src/azure-cli/azure/cli/command_modules/backup/custom_wl.py:644]: To make the message more actionable, we may also tell users which argument they can use to pass a valid workload type

File: src/azure-cli/azure/cli/command_modules/backup/custom_wl.py [hunk 3]
```diff
@@ -636,6 +639,14 @@ def _get_protected_item_instance(item_type):
...
+        if item_type_map == protectable_item_type_map:
+            raise RequiredArgumentMissingError("""
+            Protectable item type missing. Please enter a valid protectable item type.
```
Comment (houk-ms) [inline src/azure-cli/azure/cli/command_modules/backup/custom_wl.py:647]: same as above

File: src/azure-cli/azure/cli/command_modules/backup/custom_wl.py [hunk 4]
```diff
@@ -636,6 +639,14 @@ def _get_protected_item_instance(item_type):
...
+            """)
+        raise RequiredArgumentMissingError("Item type missing. Enter a valid item type.")
```
Comment (houk-ms) [inline src/azure-cli/azure/cli/command_modules/backup/custom_wl.py:649]: same as above

File: src/azure-cli/azure/cli/command_modules/backup/custom_wl.py [hunk 5]
```diff
@@ -257,6 +257,9 @@ def show_protectable_item(items, name, server_name, protectable_item_type):
 
 
 def show_protectable_instance(items, server_name, protectable_item_type):
+    if server_name is None:
+        raise RequiredArgumentMissingError("Server name missing. Please provide a valid server name.")
```
Comment (houk-ms) [inline src/azure-cli/azure/cli/command_modules/backup/custom_wl.py:261]: To make the message more actionable, we may also tell users which argument they can use to pass a valid workload type

File: src/azure-cli/azure/cli/command_modules/backup/custom_help.py [hunk 6]
```diff
@@ -305,7 +314,7 @@ def get_object_from_json(client, json_or_file, class_name):
     # Deserialize json to object
     param = client._deserialize(class_name, json_obj)  # pylint: disable=protected-access
     if param is None:
-        raise ValueError(
+        raise ResourceNotFoundError(
```
Comment (houk-ms) [inline src/azure-cli/azure/cli/command_modules/backup/custom_help.py:317]: Maybe `ValidationError` here. `ResourceNotFoundError` refers to 404 error for azure resources

File: src/azure-cli/azure/cli/command_modules/backup/custom_help.py [hunk 7]
```diff
@@ -288,7 +297,7 @@ def get_or_read_json(json_or_file):
         with open(json_or_file) as f:
             json_obj = json.load(f)
     if json_obj is None:
-        raise ValueError(
+        raise ResourceNotFoundError(
```
Comment (houk-ms) [inline src/azure-cli/azure/cli/command_modules/backup/custom_help.py:300]: Maybe ValidationError here. ResourceNotFoundError refers to 404 error for azure resources

File: src/azure-cli/azure/cli/command_modules/backup/custom_help.py [hunk 8]
```diff
@@ -108,9 +109,17 @@ def validate_policy(policy):
...
+
+
+def validate_azurefileshare_item(azurefileshare_item):
+    validate_object(azurefileshare_item, "Azure File Share item not found. Please provide a valid azure_file_share.")
+
+
 def validate_object(obj, error_message):
     if obj is None:
-        raise ValueError(error_message)
+        raise ResourceNotFoundError(error_message)
```
Comment (houk-ms) [inline src/azure-cli/azure/cli/command_modules/backup/custom_help.py:122]: Is `obj` an Azure resource? If not, maybe we should use `ValidationError` here.

File: src/azure-cli/azure/cli/command_modules/backup/custom_base.py (hunk 1 above)
Comment (houk-ms) [reply src/azure-cli/azure/cli/command_modules/backup/custom_base.py:201]: For the new error handling cases, `CLIError` is deprecated. Check the [error handling guideline here](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md)

For this case, we can use `ValidationError`

**Review by houk-ms — COMMENTED** [review]
Thanks for the effort on error improvement! Leave some comments

File: src/azure-cli/azure/cli/command_modules/backup/custom_wl.py (hunk 4 above)
Comment (akshayneema) [reply src/azure-cli/azure/cli/command_modules/backup/custom_wl.py:649]: It's just a placeholder. This case will never occur in today's code.

File: src/azure-cli/azure/cli/command_modules/backup/custom_help.py (hunk 8 above)
Comment (akshayneema) [reply src/azure-cli/azure/cli/command_modules/backup/custom_help.py:122]: Yes. obj can be a container, policy, protectable item, azure file share, etc.

File: src/azure-cli/azure/cli/command_modules/backup/custom_wl.py (hunk 5 above)
Comment (akshayneema) [reply src/azure-cli/azure/cli/command_modules/backup/custom_wl.py:261]: Actually, this function is used in other commands too so the argument can be either --server-name or --target-server-name depending on the command used.

Comment (houk-ms) [reply src/azure-cli/azure/cli/command_modules/backup/custom_wl.py:261]: Can we split the cases or pass the function an additional parameter to differentiate the arguments?

Comment (akshayneema) [reply src/azure-cli/azure/cli/command_modules/backup/custom_wl.py]: Pushed the relevant changes. Please have a look.

**Review by houk-ms — APPROVED** [review]
Error handling looks good to me.

**Review by fengzhou-msft — APPROVED** [review]

**Pull request merged**